### ISSUE TITLE
Recursive messages cause the parser to stack overflow

### ISSUE THREAD

**Issue by nsnichols** [asyncapi/parser-js#83]
#### Describe the bug
I need to define messages that are recursive tree structures. When I do this and attempt to use the `@asyncapi/generator`, to generate documentation (or code) for my spec, the parser fails with a `Error: Maximum call stack size exceeded` message.

#### How to Reproduce
Here's a minimal asyncapi spec that reproduces the issue:
```yaml
asyncapi: '2.0.0'
info:
  title: Example
  version: 0.1.1
channels:
  recursive:
    subscribe:
      message:
        payload:
          $ref: '#/components/schemas/Recursive'
components:
  schemas:
    Recursive:
      type: object
      properties:
        children:
          type: array
          items:
            $ref: '#/components/schemas/Recursive'
        # This also causes a stack overflow
        # child:
        #   $ref: '#/components/schemas/Recursive'
        something:
          type: string
```
```
$ npx -p @asyncapi/generator ag ./recursive.yaml @asyncapi/html-template --force-write -o docs
npx: installed 213 in 8.774s
Something went wrong:
Error: Maximum call stack size exceeded
    at parse (~/.npm/_npx/32509/lib/node_modules/@asyncapi/generator/node_modules/@asyncapi/parser/lib/parser.js:121:11)
    at async Generator.generateFromString (~/.npm/_npx/32509/lib/node_modules/@asyncapi/generator/lib/generator.js:233:21)
    at async /~/.npm/_npx/32509/lib/node_modules/@asyncapi/generator/cli.js:150:9
```

#### Expected behavior
I would expect the parser to be able to handle recursive messages.

**Comment by nsnichols** [comment]
I see the parser allows you to pass `{ dereference: { circular: true } }` which does stop the infinite recursion. So maybe this is a problem with the generator and not the parser. From what I can tell there isn't a way to pass that config to the parser when you run the default generator script. Though, I was able to create a generator myself and pass that config along.

```js
const Generator = require('./lib/generator');
const path = require('path');

const generator = new Generator('@asyncapi/html-template', path.resolve(__dirname, 'example'), { forceWrite: true });

(async () => {
await generator.generateFromString(`
  asyncapi: '2.0.0'
  info:
    title: Example
    version: '0.1.0'
  channels:
    example-channel:
      subscribe:
        message:
          payload:
            $ref: '#/components/schemas/Recursive'

  components:
    schemas:
      Recursive:
        type: object
        properties:
          children:
            type: array
            items:
              $ref: '#/components/schemas/Recursive'
          something:
            type: string
`, { dereference: { circular: 'ignore' } });
}) ();
```

This produces the following error:
```
(node:74327) UnhandledPromiseRejectionWarning: Template render error: (~/dev/stuff/generator/node_modules/@asyncapi/html-template/template/index.html)
  Template render error: (~/dev/stuff/generator/node_modules/@asyncapi/html-template/partials/content.html)
  Template render error: (~/dev/stuff/generator/node_modules/@asyncapi/html-template/partials/operations.html)
  Error: Your schema contains $ref. You must provide specification in the third parameter.
    at Object._prettifyError (~/dev/stuff/generator/node_modules/nunjucks/src/lib.js:36:11)
    at ~/dev/stuff/generator/node_modules/nunjucks/src/environment.js:561:19
    at eval (eval at _compile (~/dev/stuff/generator/node_modules/nunjucks/src/environment.js:631:18), <anonymous>:51:11)
    at ~/dev/stuff/generator/node_modules/nunjucks/src/environment.js:569:11
    at eval (eval at _compile (~/dev/stuff/generator/node_modules/nunjucks/src/environment.js:631:18), <anonymous>:66:12)
    at ~/dev/stuff/generator/node_modules/nunjucks/src/environment.js:569:11
    at eval (eval at _compile (~/dev/stuff/generator/node_modules/nunjucks/src/environment.js:631:18), <anonymous>:12:11)
    at /Users/j1014702/dev/stuff/generator/node_modules/nunjucks/src/environment.js:611:9
    at eval (eval at _compile (~/dev/stuff/generator/node_modules/nunjucks/src/environment.js:631:18), <anonymous>:56:12)
    at ~/dev/stuff/generator/node_modules/nunjucks/src/environment.js:611:9
```
I saw similar errors with other templates.

Should the parser be producing different output? Should the generator handle untouched `$ref`s differently? Having the parser default to `{ dereference: { circular: true } }` seems bad as this is guaranteed to cause infinite recursion when recursive schemas are present.

**Comment by derberg** [comment]
Parser uses https://apitools.dev/json-schema-ref-parser/docs/options.html and throws errors as the ref parser, so if you set `ignore` they are ignored, if you set `false` you get an error. In logs you do not see any output from parser as it just ignored refs. But then template failed as the refs stayed in the AsyncAPIDocument. I don't see any other option than having `circular: true` by default.

in case you have circular refs, what do you expect to see in docs for such a schema?

**Comment by nsnichols** [comment]
If I drop the same recursive schema into https://editor.swagger.io/, I see this:

<img width="513" alt="recursive" src="https://user-images.githubusercontent.com/25645854/83572913-0a4ac500-a4f0-11ea-8e3c-092095691e70.png">

Which isn't great, but it's something. If the non-recursive properties are shown normally and there is some indication that recursive properties are recursive, I think that's good enough.

**Comment by derberg** [comment]
@nsnichols I created a simple project and played a bit with ref resolver here https://codesandbox.io/s/parser-ref-bundle-7ogrn

When using ref resolver directly all looks like refs are resolved fine, I only had a problem with `JSON.stringify()` as it doesn't support circular deps so I had to add some library to help out. 
When using parser, it fails with `Maximum call stack size exceeded`

so basically I need to dig more why this is happening.

**Comment by nsnichols** [comment]
Thanks for looking into this.

It seems worth considering making `$ref`s a part of the parsed api output. There needs to be some way to represent a recursive structure (unless you don't plan on supporting them at all) and `$ref` does that already. I don't think you would want to make the generator and/or templates have to figure out themselves if there is recursion happening by tracking if they've seen an object already or not.

You could ensure that all non-recursive `$ref`s are resolved and everything is in a single object by calling `bundle` and then `dereference` with `circular: 'ignore'`. If you don't want `$ref` to be part of the api output, you could add a post processing step here to transform the recursive references into something else.

**Comment by derberg** [comment]
It's hard for me to imagine a use case that someone wants to have fully dereferenced circular refs, but also I don't like leaving those in the doc too. I personally like what for example https://github.com/moll/json-stringify-safe is doing and marking circular dependencies.

After deeper investigation I see parser is failing for 2 reasons:
- ajv validation should be done before dereferencing
- we have a function that recursively goes through schemas a creates anonymous ids for schema without id. Obviously it is failing in infinite loop

I think we should simply add a kind of helper that is triggered once `circular: true`, and the helper would work similar to the package I mentioned above.

@fmvilas what do you think about this one?

**Comment by fmvilas** [comment]
A few notes on this:

1. If we move ajv validation before deref, we'll have to do ajv validation again after deref.
2. The [json-schema-ref-parser package also supports this [circular] thing](https://apitools.dev/json-schema-ref-parser/docs/options.html). It used to be set up as `circular: true` but got lost at some point. We should make it the default and write some tests.
3. With 2. in place, our function should not end up in an infinite loop but I'm not 100% sure so we better write some tests there. We still may have to deref the message.payload branch when the schemaFormat is not the default one (i.e., it's RAML, OpenAPI, etc.) but that's probably a job of the schema parser itself, not this parser.

**Comment by derberg** [comment]
@fmvilas 

1. Ajv doesn't like circular refs https://github.com/ajv-validator/ajv#circular-references-in-javascript-objects so we 
2. `circular: true` is default, at least it looks like this in my project and can be indirectly assumed from the docs https://apitools.dev/json-schema-ref-parser/docs/#circular-refs. I do see `[Circular]` when I read an object, but then when I try to read if further, it goes on and on. This is super confusing, I think I lack some knowledge or something... Have a look at https://codesandbox.io/s/parser-ref-bundle-7ogrn?file=/src/index.js I tried to go through existing issues but no luck. At the end if we have no ideas I can create an inquiry issue there with my project to get some more explanation
3. I agree with what you wrote about `message.payload`.

**Comment by fmvilas** [comment]
I think you have to use `circular: 'ignore'`. This works for me:

```js
const circSchema = {
  asyncapi: '2.0.0',
  info: {
    title: 'test',
    version: '2.0.0',
  },
  channels: {},
  components: {
    schemas: {
      Recursive: {
        type: "object",
        properties: {
          children: {
            type: "array",
            items: {
              $ref: "#/components/schemas/Recursive"
            }
          },
          something: {
            type: "string"
          }
        }
      }
    }
  }
};

const $RefParser = require("@apidevtools/json-schema-ref-parser");
const stringify = require("json-stringify-safe");
const Ajv = require('ajv');
const asyncapi = require('@asyncapi/specs');

$RefParser
  .dereference(circSchema, {
    dereference: {
      circular: 'ignore'
    }
  })
  .then(schema => {
    const ajv = new Ajv({
      jsonPointers: true,
      allErrors: true,
      schemaId: 'id',
      logger: false,
    });

    const validate = ajv.compile(asyncapi['2.0.0']);
    const valid = validate(JSON.parse(stringify(schema, null, 4)));
    console.log('Valid?', valid);
    console.log(validate.errors);
  })
  .catch(console.error);
```

**Comment by derberg** [comment]
@fmvilas but `ignore` leaves `$ref` and then templates fail as they are not prepared for not dereferenced properties. 

I saw somewhere in swagger UI I think that they dereference with info that given property is circular, I believe this is the best UX really

**Comment by derberg** [comment]
we discussed it yesterday during public meeting (once recording is public, link will be here https://github.com/asyncapi/asyncapi/issues/394). 

I will try to work on it and mimic what https://github.com/moll/json-stringify-safe does. So once circular ref is discovered we stop resolving and provide information that the field has circular ref, and this info has to be easy to parse in [the generator](https://github.com/asyncapi/generator/) after all

**Comment by derberg** [comment]
@nsnichols  would be awesome if you could have a look at #94 and proposed way of handling circular refs.

tl;dr in case of circular ref we would return this info `'$.components.schemas.RecursiveSelf'` so user can figure out from this information that it is circular ref and where it is pointing.

It is hard to come up with something more meaningful without real-world use case but this PR is making a good foundation for future improvements imho

**Comment by derberg** [comment]
Another round here. Thanks @nsnichols for your patience.

`payload: '$.components.schemas.RecursiveSelf'` was a completely bad idea and at the moment it will not pass the validation of schema against AsyncAPI schema format. Also `smart-circular` is not that smart after all as it is buggy and even non-circular refs are marked as circular (thanks @nsnichols for good example, I will make sure tests cover that use case now)

so going into proposal phase again. some logic from current implementation is good and will be useful, we know exactly when we have circular refs and when to trigger dereferencing. Just a matter of deciding what to do with this info.

>  just leaving the ref alone ("key": { "$ref": "#/components/schema/SomeName" } )

we can leave it alone but it will look like this [in case of references to external files from the test from the PR](https://github.com/asyncapi/parser-js/pull/94/files#diff-24324634cc7cb5ec6be1245a12fcf4f4R15):
```
"./good/circular-refs-file-ref.yaml#/ExternalFile"
```
what I liked in the PR was that it tried to resolve as much so I get as deep as possible `parsedJSON.channels['external/file'].publish.message.payload.properties.test.properties.children.items` and then just at the end I get info about the circular ref `$.channels.external/file.publish.message.payload`. So I guess `#/channels/external/file/publish/message/payload` would be better json pointer style.

The thing here is that I would love to have it seamless for the template/generator developers so they don't need to do anything special in code for circular refs... so $ref prop is not an option but something else would have to be there I believe... 🤔

**Comment by derberg** [comment]
Final proposal (when looking at [example from PR](https://github.com/asyncapi/parser-js/pull/94/files#diff-24324634cc7cb5ec6be1245a12fcf4f4)):

This:
```
"channels": {
        "external/file": {
            "publish": {
                "message": {
                    "payload": {
                        "$ref": "./good/circular-refs-file-ref.yaml#/ExternalFile"
                    }
                }
            }
        }
}
```

Dereferences to:
```
"channels": {
  "external/file": {
            "publish": {
                "message": {
                    "payload": {
                        "type": "object",
                        "properties": {
                            "test": {
                                "type": "object",
                                "properties": {
                                    "children": {
                                        "type": "array",
                                        "items": {
                                             "$ref": "#/channels/external/file/publish/message/payload"
                                        }
                                    }
                                }
                            }
                        }
                    }
                }
            }
        }
}
```

I'm not 100%  sure it is better than just leaving 
```
"payload": {
                        "$ref": "./good/circular-refs-file-ref.yaml#/ExternalFile"
                    }
```
But this is the only best way in my opinion.

Why?

- we do not need to mark anything as `Circular` cause if $ref will stay there it means it was not resolved. But I don't have a strong opinion here, we can also add `"circular": true` property and then even on a model level have `isCircular` function, but is it really needed? I believe the developer should not be bothered to think if something is circular or not.
- We can then solve `Error: Your schema contains $ref. You must provide specification in the third parameter.` which is just about fixing https://github.com/asyncapi/generator-filters/blob/master/src/customFilters.js#L75 that needs a schema so it can dereference circular refs like this https://github.com/Redocly/openapi-sampler/blob/a5ded39229e7ae04b5c0a8c44dfc84107af1886d/test/integration.spec.js#L532. We would not be able to solve it if the $ref points to external file.

**Comment by fmvilas** [comment]
You're doing a great job here, Łukasz but let me point you to a few things:

1. Do we need to work at this level? At the stringified version level? Developers want to use the parser to work with JS structures/models instead of the stringified version of them. JS models can have circular references and we can flag them as "be careful, this is circular and it's going to start the cycle again here". That could translate to `isCircular()` (as you said) and a another method like `circularReference()` which will give you the referenced model where the cycle starts again.
1. `I believe the developer should not be bothered to think if something is circular or not.`: Agree, but if I understand correctly, your proposal is then going to leave the responsibility of resolving this hanging `$ref` to the developer, which is much worst because they'll have to parse JSON Pointers. I think we can give them this information already resolved. Also, as a developer I'd need to know sometimes if it's circular or not, to know when to stop. For instance, the HTML documentation renderer needs to know if it's circular so it stops rendering deeper levels on the UI and maybe provide a link to the initial object or simply provide a different UI.

### My proposal

1. Change [this](https://github.com/asyncapi/parser-js/blob/master/lib/parser.js#L78) to `const refParser = new $RefParser();`.
1. Bundle the document: https://apitools.dev/json-schema-ref-parser/docs/ref-parser.html#bundlepath-options-callback.
1. Try to dereference all the $ref's using `refParser.dereference()` instead of `$RefParser.dereference()`. Make sure option `dereference.circular` set to `ignore`. Not sure if this step and the previous one can be merged.
1. If [`refParser.$refs.circular === true`](https://apitools.dev/json-schema-ref-parser/docs/refs.html#circular) we check for the remaining $ref's. We now know they are circular for sure otherwise, they would have been dereferenced in step 3.
1. We add `isCircular()` and `circularReference()` to [Schema model](https://github.com/asyncapi/parser-js/blob/master/lib/models/schema.js) . The latter returns the initial schema, where cycle starts.
1. We make the Schema model work differently internally depending if it's circular or not. If it's not, it works as it's doing now, accessing the `this._json` object. If it's circular, it doesn't access the `this._json` property but instead it accesses the `circularReference()._json` object.


It's a complex topic and it's freaking difficult to explain what I want to say. I hope it makes sense 😅

**Comment by derberg** [comment]
>Do we need to work at this level? At the stringified version level?

I'm not sure what you mean? you mean my examples? I did it this way to clearly visualize how it looks like before and after

> "be careful, this is circular and it's going to start the cycle again here"

as I said, not a strong opinion from my side, I can add `isCircular()`  and `circularReference()` too

>  I think we can give them this information already resolved

Yeah, this is just a different approach, also fine. For me both approaches are qually usefull, and this is why propably ref parser enables you to decide if if you want to resolve circ refs or not. 
We just need to pick one and then handle it properly in our function where we're adding anonymous schema id.
I'm fine with resolving refs instead of leaving them.

---

The PR you looked into is old, I'm already using Bundle locally, but good point about `$refs.circular` as I haven't noticed it in the docs

**Comment by fmvilas** [comment]
> I'm not sure what you mean? you mean my examples? I did it this way to clearly visualize how it looks like before and after

That's my point, IMHO you don't have to work at AsyncAPI level but at Parser models level. Developers are going to interact with the parser not with the spec directly so you can add whatever information there, not necessarily compliant with AsyncAPI spec. We don't have to worry about how it would look like if we serialize it back to a string.

> For me both approaches are qually usefull,

Correct me if I'm wrong but there's a difference between leaving the developer with the responsibility to check if `isCircular()` and the responsibility to resolve JSON Pointers and convert them to Schema model instances. Am I missing something here?

> I'm fine with resolving refs instead of leaving them.

Yeah, we can resolve them. If then we want to serialize the whole document back to a string, we leave the `$ref` there. That's pretty ok for me because we don't have another option 😄 We could use YAML references but JSON doesn't provide a similar mechanism.

**Comment by derberg** [comment]
> We don't have to worry about how it would look like if we serialize it back to a string.

It is just how I work, I'm a `picture` person, have to visualize. Wasn't aware it will cause confusion. I look at the problem in a comples way, from html template through generator at a parser, so I just need to visualize 😄 

> Correct me if I'm wrong but there's a difference between leaving the developer with the responsibility to check

I don't want to continue the argument, I have my opinion here, but it is not a strong opinion so I will just do what was suggested.

---
Current implementation: 
- no need to use  `bundle`
- I discover circular refs where we crawl through schemas for different reasons (like assigning anonymous ids). Thanks to it parser still works even with fully dereferenced circular refs
- when circular ref is noticed, `x-parser-circular = true` is added
- also entire document, on the root level gets `x-parser-circular = true` so later you can do doc.hasCircular() to see if the AsyncAPI document you are dealing with will cause you a `headache`. Will be useful for tooling providers, that do not want to support circular refs, to easily check and throw info "circular refs are not supported"

TODO:
- Add `isCircular` on a schema level
- Fix https://github.com/asyncapi/generator-filters/blob/master/src/customFilters.js#L75 to pass entire schema to support openapi sampler functionality for circular refs

**Comment by derberg** [comment]
@fmvilas 

ok, so in pr, readme describes current behaviour https://github.com/asyncapi/parser-js/pull/94/files#diff-04c6e90faac2675aa89e2176d2eec7d8R133

I don't think `circularReference` is needed. Unless I got it wrong. But with the current implementation, you can easily access recursive data, but you can stop by checking `x-parser-circular` property.

**Comment by derberg** [comment]
some clarification on this one:
- main pull request will not solve this issue, https://github.com/asyncapi/parser-js/pull/94 is only adding nice features to detect cycyles and makes sure `parser` is never going to fail because of circular refs
- I already created PR in here https://github.com/asyncapi/html-template/pull/48 to have `html-template` support circular refs. This also required contribution to our library with generic filters, where we have a filter/function generating examples https://github.com/asyncapi/generator-filters/pull/7. The thing is that `openapi-sampler` doesn't support our use case, it supports circular refs pretty nice but only in a case when it has to resolve `$ref` on its own. I created this issue https://github.com/Redocly/openapi-sampler/issues/113 to get the opinion of the maintainer on this subject.

**Comment by derberg** [comment]
Reopen because of my [comment](https://github.com/asyncapi/parser-js/issues/83#issuecomment-659958312). Let us wait what will be the response under https://github.com/Redocly/openapi-sampler/issues/113 to make sure no more things to do in the parser and rest is only in other repos

**Comment by fmvilas** [comment]
I think I'll create a new package anyway because we don't use OpenAPI schemas but JSON Schemas, slightly different.

**Comment by derberg** [comment]
you mean like `asyncapi-sampler` or actually `jsonschema-sampler`?

**Comment by fmvilas** [comment]
`jsonschema-sampler` so it's also valid for other people/projects.

**Comment by derberg** [comment]
I created a PR to `openapi-sampler` to get support for circular refs when we generate examples -> https://github.com/Redocly/openapi-sampler/pull/115

It is used here -> https://github.com/asyncapi/html-template/pull/48 and as you can see it looks pretty good

`markdown-template` is not part of this issue but it doesn't make sense to add circular ref support to `html-template` especially if we use `openapi-sampler` there as well. So I will jump into it too.

**Comment by derberg** [comment]
work on markdown template showed me that what I have in html template is very bad, thus created another PR to the parser with better circ support https://github.com/asyncapi/parser-js/pull/161

**Comment by derberg** [comment]
@nsnichols hey, this 👇 is how the rendering of circular refs will look like in markdown and html templates once we merge all my PRs. What do you think?

<img width="1621" alt="Screenshot 2020-08-13 at 18 52 38" src="https://user-images.githubusercontent.com/6995927/90222465-62fdde80-de0c-11ea-979f-a4d4c687e848.png">
<img width="858" alt="Screenshot 2020-08-13 at 17 54 22" src="https://user-images.githubusercontent.com/6995927/90222473-6729fc00-de0c-11ea-8d67-b0d1b8b9599d.png">

**Comment by nsnichols** [comment]
@derberg, thanks for all the work you’ve done on this! The new output looks good.

One thing that is a little confusing to me in the example data is children showing up as both an array and an object. I’m not sure what you’re actually using as the schema so maybe it’s correct, but at first glance, the example is confusing to me. The way it’s _rendered_ looks nice.

**Comment by derberg** [comment]
@nsnichols no worries. Sorry for the confusion, it is because when I worked on more complex example and did some copy/paste I did not change `children` to something else, so in one schema it is array and in another it is object:

```
asyncapi: '2.0.0'
info:
  title: Example
  version: 0.1.1
channels:
  recursive:
    subscribe:
      message:
        payload:
          $ref: '#/components/schemas/Recursive'
components:
  schemas:
    Recursive:
      type: object
      properties:
        children:
          type: array
          items:
            $ref: '#/components/schemas/Recursive'
        something:
          type: object
          properties:
            maryna:
              $ref: '#/components/schemas/RecursiveNext'
    RecursiveNext:
      type: object
      properties:
        children:
          type: object
          properties:
            maniek:
              $ref: '#/components/schemas/RecursiveNext'
            ulala:
              type: string
```

We will wait a week or 2 for the merge of my PR to openapi-sampler and once it is done, the PRs to html and markdown templates will be merged

**Comment by derberg** [comment]
@nsnichols playground is updated to use latest parser, generator and html/markdown templates -> all with support for circular refs. Give it a try and create a followup issue in case you think it needs more improvements
